components/script/dom/xrsession.rs

```diff
@@ -76,30 +146,93 @@ impl XRSessionMethods for XRSession {
-            self.base_layer.get().as_ref().map(|l| &**l),
-        )
+        self.active_render_state.get()
     }
 
-    /// https://immersive-web.github.io/webxr/#dom-xrsession-requestanimationframe
-    fn UpdateRenderState(&self, init: &XRRenderStateInit, comp: InCompartment) -> Rc<Promise> {
-        let p = Promise::new_in_current_compartment(&self.global(), comp);
-        self.display.queue_renderstate(init, p.clone());
-        p
+    /// https://immersive-web.github.io/webxr/#dom-xrsession-updaterenderstate
+    fn UpdateRenderState(&self, init: &XRRenderStateInit, _: InCompartment) {
+        // XXXManishearth various checks:
+        // If session’s ended value is true, throw an InvalidStateError and abort these steps
+        // If newState’s baseLayer's was created with an XRSession other than session,
+        // throw an InvalidStateError and abort these steps
+        // If newState’s inlineVerticalFieldOfView is set and session is an
+        // immersive session, throw an InvalidStateError and abort these steps.
+
+        let pending = self
+            .pending_render_state
+            .or_init(|| self.active_render_state.get().copy());
+        if let Some(near) = init.depthNear {
+            pending.set_depth_near(*near);
+        }
+        if let Some(far) = init.depthFar {
+            pending.set_depth_far(*far);
+        }
+        if let Some(ref layer) = init.baseLayer {
+            pending.set_layer(Some(&layer))
+        }
+        // XXXManishearth handle inlineVerticalFieldOfView
     }
 
     /// https://immersive-web.github.io/webxr/#dom-xrsession-requestanimationframe
     fn RequestAnimationFrame(&self, callback: Rc<XRFrameRequestCallback>) -> i32 {
-        self.display.xr_raf(callback) as i32
+        #[derive(serde::Serialize, serde::Deserialize)]
+        pub struct FrameCallback {
+            sender: IpcSender<(f64, Frame)>,
+        }
+
+        #[typetag::serde]
+        impl webxr_api::FrameRequestCallback for FrameCallback {
+            fn callback(&mut self, time: f64, frame: Frame) {
+                let _ = self.sender.send((time, frame));
+            }
+        }
+
+        // queue up RAF callback, obtain ID
+        let raf_id = self.next_raf_id.get();
+        self.next_raf_id.set(raf_id + 1);
+        self.raf_callback_list
+            .borrow_mut()
+            .push((raf_id, Some(callback)));
+
+        // set up listener for response, if necessary
+        if self.raf_sender.borrow().is_none() {
+            let this = Trusted::new(self);
+            let global = self.global();
+            let window = global.as_window();
+            let (task_source, canceller) = window
+                .task_manager()
+                .dom_manipulation_task_source_with_canceller();
+            let (sender, receiver) = ipc::channel(global.time_profiler_chan().clone()).unwrap();
+            *self.raf_sender.borrow_mut() = Some(sender);
+            ROUTER.add_route(
```

Comment:
We can discuss this in #23733.

Reply:
Ok so, if the goal is really removing the router from the equation for this specific workflow(not the event-loop as a whole, please), then I have two potential solutions to propose for consideration:

1. The "well-behaved" variant: Somewhere at the end of [`ScriptThread.handle_msg`](https://github.com/servo/servo/blob/d0bd2d5e44d4006b00ef53f351945d01e1c8458d/components/script/script_thread.rs#L1457), check for the presence of an active Xr session, and of animation frame callbacks, and if there are any callbacks, do a blocking `recv()` on the raf ipc-receiver, until you've ran all the callbacks(can you count on the frames being provided is asked for?). Alternatively, do a non-blocking `try_recv`, maybe spin a few times to try if any frame is received, and if not just continue with the rest of the event-loop. 
2. The "still well-behaved, but somewhat more aggressive" variant: Each time you run a task, after you've done a microtask checkpoint [here](https://github.com/servo/servo/blob/d0bd2d5e44d4006b00ef53f351945d01e1c8458d/components/script/script_thread.rs#L1443), do a `try_recv()` on the raf ipc receiver. Combine this with a blocking `recv` as described under 1. 
3. The "experimental" variant: At this very line, instead of using the `ROUTER`, queue a task that does only one thing: blockingly `recv()` on the raf ipc receiver, until you've either run out of callbacks, or the Xr session has ended. You might want to `select` on a second ipc receiver, say, something that would tell you when the user made a gesture indicating the wish to end the session immediately. 

General background on this idea: the concept of a [Frame loop](https://immersive-web.github.io/webxr/#frame) doesn't seem to necessarily imply running other tasks enqueued on the event-loop, also it doesn't say anything about queuing a task to run an animation frame callback. So one could basically run a "frame loop" as some sort of tight-loop on a ipc receiver, while stopping the rest of the event-loop for the page(so long as you still allow the user to end the session).

Reply:
One last thing: in all cases, before starting the "Frame loop", make sure to do a `self.background_hang_monitor.notify_wait();` on the `ScriptThread`, otherwise it will look like you're hanging and it will suspend the script thread and sample it's registers!

Reply:
On a more general note: it's not completely clear to me how the Xr "frame loop" is supposed to integrate with the window event-loop, if at all. 

If you don't need to actually run tasks on the window event-loop, or handle most messages from the constellation, you could try an even more drastic approach where instead of running one callback in one task, you'd keep requesting frames and running callbacks for as long as new callbacks are added inside a callback, and for as long as the session remains active(and the pipeline isn't closed?). And that from within the same task, effectively blocking the script-thread in the meantime(but running the callbacks as "fast" as possible, I guess).

Reply:
We could do something like that, but this would mean blocking the script thread while waiting for the `raf_receiver.recv()`, and IPC channels don't support `recv_timeout()`. Sigh. I think this would need quite a drastic re-think of how Servo approaches IPC.